tiledb/sm/serialization/array_schema.cc

```diff
@@ -304,110 +306,186 @@ Status dimension_to_capnp(
+    case Datatype::DATETIME_MIN:
+    case Datatype::DATETIME_SEC:
+    case Datatype::DATETIME_MS:
+    case Datatype::DATETIME_US:
+    case Datatype::DATETIME_NS:
+    case Datatype::DATETIME_PS:
+    case Datatype::DATETIME_FS:
+    case Datatype::DATETIME_AS:
+    case Datatype::TIME_HR:
+    case Datatype::TIME_MIN:
+    case Datatype::TIME_SEC:
+    case Datatype::TIME_MS:
+    case Datatype::TIME_US:
+    case Datatype::TIME_NS:
+    case Datatype::TIME_PS:
+    case Datatype::TIME_FS:
+    case Datatype::TIME_AS:
+    case Datatype::INT64: {
+      auto val = tile_extent_reader.getInt64();
+      std::memcpy(tile_extent.data(), &val, coord_size);
+      break;
+    }
+    case Datatype::UINT64: {
+      auto val = tile_extent_reader.getUint64();
+      std::memcpy(tile_extent.data(), &val, coord_size);
+      break;
+    }
+    case Datatype::FLOAT32: {
+      auto val = tile_extent_reader.getFloat32();
+      std::memcpy(tile_extent.data(), &val, coord_size);
+      break;
+    }
+    case Datatype::FLOAT64: {
+      auto val = tile_extent_reader.getFloat64();
+      std::memcpy(tile_extent.data(), &val, coord_size);
+      break;
+    }
+    default:
+      throw std::logic_error(
+          "[Deserialization::tile_extent_from_capnp] Failed to deserialize "
+          "tile extent from capnp.");
+  }
+  return tile_extent;
+}
 
+/** Deserialize a range from a cap'n proto object.
+ *
+ * @pre dim_type is valid.
+ */
+Range range_from_capnp(
+    Datatype dim_type, const capnp::Dimension::Reader& dimension_reader) {
   if (dimension_reader.hasDomain()) {
     auto domain_reader = dimension_reader.getDomain();
     Buffer domain_buffer;
-    RETURN_NOT_OK(
-        utils::copy_capnp_list(domain_reader, dim_type, &domain_buffer));
-    RETURN_NOT_OK((*dimension)->set_domain_unsafe(domain_buffer.data()));
+    Status st = utils::copy_capnp_list(domain_reader, dim_type, &domain_buffer);
+    if (!st.ok()) {
+      throw std::runtime_error(
+          "[Deserialization::range_from_capnp] Failed to deserialize domain.");
+    }
+    return Range{domain_buffer.data(), datatype_size(dim_type) * 2};
+  } else {
+    return Range{};
   }
+}
+
+/** Deserialize a dimension from a cap'n proto object. */
+shared_ptr<Dimension> dimension_from_capnp(
+    const capnp::Dimension::Reader& dimension_reader) {
+  Status st;
+
+  // Deserialize datatype
+  Datatype dim_type;
+  st = datatype_enum(dimension_reader.getType().cStr(), &dim_type);
+  if (!st.ok()) {
+    throw std::runtime_error(
+        "[Deserialization::dimension_from_capnp] " +
+        std::string(dimension_reader.getType().cStr()) +
+        " is not a valid datatype identifer.");
+  }
+
+  // Validate the dim_type, satisfying precondition of tile_extent_from_capnp
+  try {
+    ensure_datatype_is_valid(dim_type);
+  } catch (const std::exception& e) {
+    std::throw_with_nested(std::runtime_error(
+        "[Deserialization::dimension_from_capnp] " + std::string(e.what())));
```

Comment:
You don't need to call `e.what()` with `throw_with_nested`; the current exception (here it's `e`) is what's nested.

Reply:
Clarified offline; `e.what()` removed